pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go

```diff
@@ -0,0 +1,523 @@
+		return nil, err
+	}
+	return items[(len(items)-1)/2], nil
+}
+
+func sortAsDecimal(items []values.Value) error {
+	var err error
+	sort.Slice(items, func(i, j int) bool {
+		decimalI, errI := toDecimal(items[i])
+		if errI != nil {
+			err = errI
+		}
+		decimalJ, errJ := toDecimal(items[j])
+		if errJ != nil {
+			err = errJ
+		}
+		return decimalI.GreaterThan(decimalJ)
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func toDecimal(item values.Value) (decimal.Decimal, error) {
+	unwrapped, err := item.Unwrap()
+	if err != nil {
+		return decimal.NewFromInt(0), err
+	}
+
+	switch v := unwrapped.(type) {
+	case string:
+		deci, err := decimal.NewFromString(unwrapped.(string))
+		if err != nil {
+			return decimal.NewFromInt(0), err
+		}
+		return deci, nil
+	case decimal.Decimal:
+		return unwrapped.(decimal.Decimal), nil
+	case int64:
+		return decimal.NewFromInt(unwrapped.(int64)), nil
+	case *big.Int:
+		big := unwrapped.(*big.Int)
+		return decimal.NewFromBigInt(big, 10), nil
+	case time.Time:
+		return decimal.NewFromInt(unwrapped.(time.Time).Unix()), nil
+	case float64:
+		return decimal.NewFromFloat(unwrapped.(float64)), nil
+	default:
+		// unsupported type
+		return decimal.NewFromInt(0), fmt.Errorf("unable to convert type %T to decimal", v)
+	}
+}
+
+func mode(items []values.Value) (values.Value, error) {
+	if len(items) == 0 {
+		// invariant, as long as f > 0 there should be items
+		return nil, errors.New("items cannot be empty")
+	}
+
+	counts := make(map[[32]byte]*counter)
+	for _, item := range items {
+		marshalled, err := proto.MarshalOptions{Deterministic: true}.Marshal(values.Proto(item))
+		if err != nil {
+			// invariant: values should always be able to be proto marshalled
+			return nil, err
+		}
+		sha := sha256.Sum256(marshalled)
+		elem, ok := counts[sha]
+		if !ok {
+			counts[sha] = &counter{
+				fullObservation: item,
+				count:           1,
+			}
+		} else {
+			elem.count++
+		}
+	}
+
+	var maxCount int
+	for _, ctr := range counts {
+		if ctr.count > maxCount {
+			maxCount = ctr.count
+		}
+	}
+
+	var modes []values.Value
+	for _, ctr := range counts {
+		if ctr.count == maxCount {
+			modes = append(modes, ctr.fullObservation)
+		}
+	}
+
+	// If more than one mode found, choose first
+
+	return modes[0], nil
+}
+
+func deviation(method string, previousValue values.Value, nextValue values.Value) (decimal.Decimal, error) {
+	prevDeci, err := toDecimal(previousValue)
+	if err != nil {
+		return decimal.NewFromInt(0), err
+	}
+	nextDeci, err := toDecimal(nextValue)
+	if err != nil {
+		return decimal.NewFromInt(0), err
+	}
+
+	diff := prevDeci.Sub(nextDeci).Abs()
+
+	switch method {
+	case DEVIATION_TYPE_ABSOLUTE:
+		return diff, nil
+	case DEVIATION_TYPE_PERCENT:
+		if prevDeci.Cmp(decimal.NewFromInt(0)) == 0 {
+			if diff.Cmp(decimal.NewFromInt(0)) == 0 {
+				return decimal.NewFromInt(0), nil
+			}
+			return decimal.NewFromInt(math.MaxInt), nil
+		}
+		return diff.Div(prevDeci), nil
```

Comment:
should there be a zero check here or is it not possible for prevDeci to be 0?

Reply:
The zero check is right above there :grin: 

```
		if prevDeci.Cmp(decimal.NewFromInt(0)) == 0 {
			if diff.Cmp(decimal.NewFromInt(0)) == 0 {
				return decimal.NewFromInt(0), nil
			}
			return decimal.NewFromInt(math.MaxInt), nil
		}
```